### PR TITLE
ci: test with Python 3.13

### PR DESCRIPTION
**Pull request by brettlangdon**
This change adjusts CI and the library itself to work under Python 3.13. Any tests that failed under 3.13 are skipped on 3.13 for now and will be unskipped in a future change.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Comment by emmettbutler** [comment]
This is blocked by https://github.com/MatthieuDartiailh/bytecode/pull/146

File: .gitlab/package.yml
Comment (P403n1x87) [inline .gitlab/package.yml:1]: I think these gitlab sources need to be aligned with what we have on the main branch now

File: ddtrace/internal/_threads.cpp
Comment (P403n1x87) [inline ddtrace/internal/_threads.cpp:23]: we normally write this with `>=`

File: tests/internal/test_forksafe.py
Comment (P403n1x87) [inline tests/internal/test_forksafe.py:303]: do we know why this fails?

File: tests/internal/test_wrapping.py
Comment (P403n1x87) [inline tests/internal/test_wrapping.py:98]: We can't skip these I'm afraid. If we have segfaults it means we don't support 3.13 and need to fix (also the condition seems wrong 🤔 )

File: ddtrace/appsec/_common_module_patches.py
Comment (brettlangdon) [inline ddtrace/appsec/_common_module_patches.py:64]: this seems out of date with `main`

File: ddtrace/debugging/_debugger.py
Comment (brettlangdon) [inline ddtrace/debugging/_debugger.py:27]: is this needed for 3.13 support?

File: tests/debugging/function/test_store.py
Comment (brettlangdon) [inline tests/debugging/function/test_store.py:178]: was this needed for 3.13 support?

**Review by brettlangdon — COMMENTED** [review]
I feel like a few things aren't in sync with main?

File: hatch.toml
Comment (gnufede) [inline hatch.toml:186]: Does this mean these don't pass with >3.10 ?

File: ddtrace/appsec/_common_module_patches.py
Comment (gnufede) [inline ddtrace/appsec/_common_module_patches.py:66]: It's indeed out of date, I think it's safe to just revert this

```suggestion
    if asm_config._iast_enabled:
```

File: ddtrace/debugging/_debugger.py
Comment (P403n1x87) [reply ddtrace/debugging/_debugger.py:27]: 🤔 this looks like a slight refactor I did. I don't think it is needed for 3.13 support so I'll see if I can take this out in its own PR

File: tests/internal/test_forksafe.py
Comment (emmettbutler) [reply tests/internal/test_forksafe.py:303]: I haven't looked deeply into it yet.

File: hatch.toml
Comment (emmettbutler) [reply hatch.toml:186]: I think this was added unintentionally

File: .gitlab/package.yml
Comment (emmettbutler) [reply .gitlab/package.yml:1]: Are you referring to these? https://github.com/DataDog/dd-trace-py/pull/10821/files/9faf26190481fb505538481d3588b22e13a2d998

File: tests/internal/test_wrapping.py
Comment (emmettbutler) [reply tests/internal/test_wrapping.py:98]:
```
tests/internal/test_wrapping.py::test_wrap_generator_yield_from Fatal Python error: Segmentation fault                                                                                                                                                                        
                                                                                                                                                                                                                                                                              
Current thread 0x0000ffffb4380020 (most recent call first):                                                                                                                                                                                                                   
  File "/root/project/tests/internal/test_wrapping.py", line 244 in g                                                                                                                                                                                                         
  File "/root/project/tests/internal/test_wrapping.py", line 248 in test_wrap_generator_yield_from                
```

Comment (emmettbutler) [reply tests/internal/test_wrapping.py:98]: This sounds maybe related https://docs.python.org/3/whatsnew/3.13.html#cpython-bytecode-changes

Comment (emmettbutler) [reply tests/internal/test_wrapping.py:98]: We decided in a different thread that these skips are ok to merge because no integrations use the functionality that these tests exercise.

File: ddtrace/appsec/_common_module_patches.py
Comment (emmettbutler) [reply ddtrace/appsec/_common_module_patches.py:66]: @gnufede this is actually a change I made to get around the fact that `_taint_tracking/__init__.py` unconditionally uses names that are conditionally imported.

File: ddtrace/debugging/_debugger.py
Comment (emmettbutler) [reply ddtrace/debugging/_debugger.py]: https://github.com/DataDog/dd-trace-py/pull/11721

**Review by gnufede — APPROVED** [review]

File: ddtrace/internal/injection.py
Comment (erikayasuda) [inline ddtrace/internal/injection.py:115]: So this is exclusive to 3.11 and 3.12?

File: ddtrace/internal/wrapping/__init__.py
Comment (erikayasuda) [inline ddtrace/internal/wrapping/__init__.py:148]: Why do we swap these for 3.13?

File: releasenotes/notes/threethirteen-d40d659d8939fe5e.yaml
Comment (erikayasuda) [inline releasenotes/notes/threethirteen-d40d659d8939fe5e.yaml:4]: nit:

```suggestion
    Makes the library compatible with Python 3.13
```

**Review by erikayasuda — APPROVED** [review]
Just some NB questions, otherwise LGTM.

I think we need to add a 3.13 image in magic mirror before merging this? Looks like the 3.13 download job is failing

File: ddtrace/internal/injection.py
Comment (emmettbutler) [reply ddtrace/internal/injection.py:115]: That's right, the opcodes changed a bunch in the last few minor versions

File: ddtrace/internal/wrapping/__init__.py
Comment (emmettbutler) [reply ddtrace/internal/wrapping/__init__.py:148]: I actually [searched](https://github.com/python/cpython/blob/f823910bbd4bf01ec3e1ab7b3cb1d77815138296/Python/flowgraph.c#L2655) for quite a while for the answer, and couldn't find it. @P403n1x87?

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/profiling/collector/stack.pyx
Comment (sanchda) [inline ddtrace/profiling/collector/stack.pyx:160]: I think this is OK

File: setup.py
Comment (sanchda) [inline setup.py:556]: Ah, right.  We're going to have to fix this.

File: tests/internal/crashtracker/test_crashtracker.py
Comment (sanchda) [inline tests/internal/crashtracker/test_crashtracker.py:509]: Very unexpected, but good to know.

**Review by sanchda — APPROVED** [review]
Looks good from profiling

File: ddtrace/internal/injection.py
Comment (P403n1x87) [inline ddtrace/internal/injection.py:115]:
```suggestion
_INJECT_HOOK_OPCODE_POS = 1 if (3,11) <= PY < (3, 13) else 0
```

File: ddtrace/internal/_threads.cpp
Comment (P403n1x87) [inline ddtrace/internal/_threads.cpp:23]:
```suggestion
#if PY_VERSION_HEX >= 0x030d0000
```
for readability 🙏

**Review by P403n1x87 — APPROVED** [review]
A few nits, otherwise LGTM

**Pull request merged**
